### PR TITLE
Add push down documentation

### PR DESCRIPTION
**Pull request by mosabua**
As discussed with @electrum and @martint . This is a first draft. Could you supply the explain plan snippets @findepi .. and maybe some more info on what functions can be pushed down or whatever else is missing.

### CODE REVIEW

File: presto-docs/src/main/sphinx/optimizer/push-down.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+Push down depends on a number of factors:
+
+* generic support for push down in the connector
+* function or operation specific support for push down in the connector
+* query that allows the detection of the function to push down
+
+The best way to analyze if push down for a specific query is performed, is to
+take a closer look at the :doc:`EXPLAIN plan </sql/explain>` of the query. If an
+operations such as an aggregate function successfully pushed down to the
+connector, the explain plan does **not** show that operator. The explain plan
```
Comment (Lewuathe) [inline presto-docs/src/main/sphinx/optimizer/push-down.rst:27]: Is that true? If the pushdown result is not guaranteed by the connector, Presto should keep the filter. For example, pushdown of limit is not guaranteed if `LimitApplicationResult.limitGuaranteed` is false. In such a case, the explain plan would show the operator while the pushdown happens in the connector.

Comment (mosabua) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst:27]: Not sure .. we need some user facing docs for this and this is my attempt to get this started.

**Comment by mosabua** [comment]
Can you please help me out here with some explain plans and other missing info @findepi ?

**Comment by findepi** [comment]
@mosabua sure! 

you can get EXPLAIN plans e.g. with `io.prestosql.plugin.postgresql.PostgreSqlQueryRunner#main`

1. run this
2. `presto-cli/target/presto-cli-*-executable.jar --debug --server localhost:8080 --catalog postgresql --schema tpch`
3. `EXPLAIN SELECT regionkey, count(*) FROM nation GROUP BY 1;`

```
presto:tpch> EXPLAIN SELECT regionkey, count(*) FROM nation GROUP BY 1;
                                                                                      Query Plan
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [regionkey_0, _presto_generated_1]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Output[regionkey, _col1]
     │   Layout: [regionkey_0:bigint, _presto_generated_1:bigint]
     │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: ?}
     │   regionkey := regionkey_0
     │   _col1 := _presto_generated_1
     └─ RemoteSource[1]
            Layout: [regionkey_0:bigint, _presto_generated_1:bigint]

 Fragment 1 [SOURCE]
     Output layout: [regionkey_0, _presto_generated_1]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     TableScan[postgresql:tpch.nation tpch.nation columns=[regionkey:bigint:int8, count(*):_presto_generated_1:bigint:bigint] groupingSets=[[regionkey:bigint:int8]], gro
         Layout: [regionkey_0:bigint, _presto_generated_1:bigint]
         Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
         _presto_generated_1 := count(*):_presto_generated_1:bigint:bigint
         regionkey_0 := regionkey:bigint:int8
```

File: presto-docs/src/main/sphinx/optimizer/push-down.rst (hunk 1 above)
Comment (mosabua) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst:27]: @findepi can you confirm what I wrote or provide corrected info?

File: presto-docs/src/main/sphinx/optimizer/push-down.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    EXPLAIN
+    SELECT regionkey, count(*)
+    FROM nation
+    WHERE regionkey = 1
+    GROUP BY 1;
+
+The explain plan shows the count operation being performed by Presto, since no
+push down with an associated condition can be performed:
+
+TBD - add explain plan
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/optimizer/push-down.rst:87]: Can you supply this as well please @findepi

File: presto-docs/src/main/sphinx/optimizer/push-down.rst [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    GROUP BY 1;
+
+You can get the explain plan with the following statement::
+
+    EXPLAIN
+    SELECT regionkey, count(*)
+    FROM nation
+    GROUP BY 1;
+
+The explain plan for this query does not show any ``count`` operation in Presto.
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/optimizer/push-down.rst:46]: Is this correct?

**Comment by mosabua** [comment]
Also please let me know if there are other aspects still missing that are in 339 or have been merged towards 340 already.

File: presto-docs/src/main/sphinx/optimizer/push-down.rst (hunk 1 above)
Comment (findepi) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst:28]: @Lewuathe this is true for _filters_ and JDBC connectors (although I am improving this in https://github.com/prestosql/presto/pull/4565)
however, there is no way to "improve aggregation", connector must do this thing correct, or not do this at all

@mosabua yes, pushded down agg won't show up in EXPLAIN anymore

File: presto-docs/src/main/sphinx/optimizer/push-down.rst (hunk 2 above)
Comment (findepi) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst:87]:
```
presto:default> EXPLAIN SELECT regionkey, count(*) FROM tpch.tiny.nation GROUP BY 1;
                                            Query Plan
---------------------------------------------------------------------------------------------------
 Fragment 0 [SINGLE]
     Output layout: [regionkey, count]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Output[regionkey, _col1]
     │   Layout: [regionkey:bigint, count:bigint]
     │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
     │   _col1 := count
     └─ RemoteSource[1]
            Layout: [regionkey:bigint, count:bigint]

 Fragment 1 [HASH]
     Output layout: [regionkey, count]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Aggregate(FINAL)[regionkey]
     │   Layout: [regionkey:bigint, count:bigint]
     │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
     │   count := count("count_0")
     └─ LocalExchange[HASH][$hashvalue] ("regionkey")
        │   Layout: [regionkey:bigint, count_0:bigint, $hashvalue:bigint]
        │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
        └─ RemoteSource[2]
               Layout: [regionkey:bigint, count_0:bigint, $hashvalue_1:bigint]

 Fragment 2 [SOURCE]
     Output layout: [regionkey, count_0, $hashvalue_2]
     Output partitioning: HASH [regionkey][$hashvalue_2]
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Project[]
     │   Layout: [regionkey:bigint, count_0:bigint, $hashvalue_2:bigint]
     │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
     │   $hashvalue_2 := combine_hash(bigint '0', COALESCE("$operator$hash_code"("regionkey"), 0))
     └─ Aggregate(PARTIAL)[regionkey]
        │   Layout: [regionkey:bigint, count_0:bigint]
        │   count_0 := count(*)
        └─ TableScan[tpch:nation:sf0.01, grouped = false]
               Layout: [regionkey:bigint]
               Estimates: {rows: 25 (225B), cpu: 225, memory: 0B, network: 0B}
               regionkey := tpch:regionkey


(1 row)
```

something like this?

Comment (mosabua) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst:87]: Not really ... in this example I talk about adding a condition so having the query as 

    EXPLAIN
    SELECT regionkey, count(*)
    FROM nation
    WHERE regionkey = 1
    GROUP BY 1;

which then should trip the push down NOT to happen.. and I want to show how that is visible in the explain plan ... your snippet above shows a different explain plan for the same query. Not sure how you generated that ..

**Comment by mosabua** [comment]
I rebased and for some reason the changes in presto-main show up here now.. not sure if github is confused or something else is wrong. I can reissue the PR off a new branch if desired or we can continue here .

File: presto-docs/src/main/sphinx/optimizer/push-down.rst (hunk 2 above)
Comment (mosabua) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst]: For now I added the above explain even though it is probably wrong at the moment.

File: presto-docs/src/main/sphinx/connector/postgresql.rst [hunk 4]
```diff
@@ -89,6 +89,18 @@ Finally, you can access the ``clicks`` table in the ``web`` schema::
 If you used a different name for your catalog properties file, use
 that catalog name instead of ``postgresql`` in the above examples.
 
+Push down
+---------
+
+The connector supports :doc:`push down </optimizer/push-down>` for aggregate
+functions:
+
+* avg
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/connector/postgresql.rst:98]: reminder .. use func ref

File: presto-docs/src/main/sphinx/optimizer/push-down.rst (hunk 2 above)
Comment (mosabua) [reply presto-docs/src/main/sphinx/optimizer/push-down.rst]: Updatd

**Comment by mosabua** [comment]
Replaced by https://github.com/prestosql/presto/pull/4660

**Pull request closed without merging**